### PR TITLE
#164694215 Receive jwt token on successfull registration

### PR DESCRIPTION
**Pull request by frankopkusianwar**
**What does this PR do?**
 
Create a  JWT token for a new registered user  

**Description of Task to be completed?**

- register new user  
- receive jwt token on successfull registration
- receive jwt token when a user sucessfully logs in
- authenticate a user to to acess restricted endpoints

**How should this be manually tested?**

**To test locally**
- clone the repo and make a virtual environment venv 
- install requirements using (pip install -r requirements.txt)
- add the environment variables to the .env file and install them(python source .env) 
- run and apply migrations (python manage.py makemigrations) then (manage.py migrate)
- finally run the server 

**postman** 
* key: content-type value: application/json
* to test authentication add this to header, (key:Authorization value: Bearer token). get token from login or signup endpoint  
* registration data {"user":{"email":"examplemail.com","username":"examplename","password":"examplepassword"}}
* user update data {"user":{"username":"updatename"}}
* enpoits post(api/users), post(api/users/login) and put(api/user)

**testing online**
- alternatively visit the review app on heroku and create a new user with email,username and password

**What are the relevant pivotal tracker stories?**

#164694215]

**Screenshots**
- login data
login data
- signup data
data format
- update user data
update data
- when there is no token in the headers
no auth token
- when the token is invalid
invalid token
- when a user successfully signs up
successfull signup token
- when the token used to access a restricted endpoint is valid
valid token

### CODE REVIEW

**Comment by Rhytah** [comment]
Great job, I have tested out the endpoint and works as it should. Now let's see the implementation of JWT

File: authors/apps/authentication/models.py [hunk 1]
```diff
@@ -112,3 +118,12 @@ def get_short_name(self):
...
         return self.username
+
+    @property
+    def auth_token(self):
+        """creates a token for registration"""
+        self.encoded_jwt = jwt.encode(
+            {'name': self.username,
+             'exp': datetime.now() + timedelta(minutes=15)},
+            SECRET_KEY, algorithm='HS256').decode('utf-8')
+        return self.encoded_jwt
```
Comment (ghost) [inline authors/apps/authentication/models.py:128]: please try using the Token from the rest_framework.authtoken.models, which is provided for you from the rest_framework

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -1,28 +1,51 @@
...
 
 from .renderers import UserJSONRenderer
 from .serializers import (
     LoginSerializer, RegistrationSerializer, UserSerializer
 )
 
 
-class RegistrationAPIView(APIView):
+class RegistrationAPIView(ListCreateAPIView):
     # Allow any user (authenticated or not) to hit this endpoint.
```
Comment (ManuelDominic) [inline authors/apps/authentication/views.py]: Please try using the proper formats for the classes with generic views

**Review by ghost — CHANGES_REQUESTED** [review]
- Please check out the proper format of the use of generic views in classes
- Try using the Token from rest_framework.authtoken.models in the rest_framework package

**Review by ghost — DISMISSED** [review]

File: authors/apps/authentication/tests/test_views.py [hunk 3]
```diff
@@ -40,3 +40,25 @@ def test_login_with_invalid_user_fails(self):
...
+        self.assertIn('token', response.data)
+        self.assertEqual(response.status_code, 200)
+
+# test whether a token is generated on signup
+    def test_signup_user_signup_token(self):
+        response = self.client.post(
+            reverse('auth:register'),
+            data=self.user_data,
+            format='json')
+        self.assertIn('generate_auth_token', response.data)
```
Comment (kimbugp) [inline authors/apps/authentication/tests/test_views.py]: For uniformity, you should use the same key identifier for the token for both login and sign up

**Review by kimbugp — CHANGES_REQUESTED** [review]

**Comment by kimbugp** [comment]
@frankopkusianwar , could you please ensure to follow the right commit convention.
Also  add more details to the PR description of how to test this including the request body etc.

**Comment by kimbugp** [comment]
@frankopkusianwar .please reference the PT board story link on the PR description

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -15,14 +17,16 @@ class RegistrationSerializer(serializers.ModelSerializer):
...
+
     # The client should not be able to send a token along with a registration
     # request. Making `token` read-only handles that for us.
 
     class Meta:
         model = User
         # List all of the fields that could possibly be included in a request
         # or response, including fields specified explicitly above.
-        fields = ['email', 'username', 'password']
+        fields = ['email', 'username', 'password', 'generate_auth_token', 'token']
```
Comment (kimbugp) [inline authors/apps/authentication/serializers.py]: @frankopkusianwar , you should have one key to represent the token

**Review by kimbugp — CHANGES_REQUESTED** [review]

**Pull request merged**